lib/pbench/server/api/resources/pbench_users.py

```diff
@@ -0,0 +1,523 @@
+            )
+        except NoOptionError:
+            self.token_expire_duration = None
+
+    def post(self):
+        """
+        Post request for registering a new user.
+        This requires a JSON data with required user fields
+        {
+            "username": "username",
+            "password": "password",
+            "firstName": first_name,
+            "lastName": "last_name",
+            "email": "user@domain.com"
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        :return: JSON Payload
+        if we succeed to add a user entry in database, the returned response_object will include the auth_token
+            response_object = {
+                    "status": "success"/"fail",
+                    "status_code": 201/status_code_int,
+                    "message": "Successfully registered."/"failure message",
+                    "auth_token": auth_token.decode(), # Will not present if failed
+                }
+        """
+        # get the post data
+        user_data = request.get_json()
+        if not user_data:
+            self.logger.error("Register User: Invalid json object: %s", request.url)
+            abort(400, message="Register User: Invalid json object in request")
+
+        username = user_data.get("username")
+        if not username:
+            self.logger.error(f"Register User: missing username field")
+            abort(
+                400, message=f"missing username field",
+            )
+        if username.lower() == "admin":
+            self.logger.error(f"User tried to register with admin username")
+            abort(
+                400, message=f"Please choose another username",
+            )
+
+        password = user_data.get("password")
+        if not password:
+            self.logger.error(f"Register User: missing password field")
+            abort(
+                400, message=f"missing password field",
+            )
+
+        emailID = user_data.get("email")
+        if not emailID:
+            self.logger.error(f"Register User: missing email field")
+            abort(
+                400, message=f"missing email field",
+            )
+
+        firstName = user_data.get("firstName")
+        if not firstName:
+            self.logger.error(f"Register User: missing firstName field")
+            abort(
+                400, message=f"missing firstName field",
+            )
+
+        lastName = user_data.get("lastName")
+        if not lastName:
+            self.logger.error(f"Register User: missing lastName field")
+            abort(
+                400, message=f"missing lastName field",
+            )
+
+        # validate the email field
+        try:
+            valid = validate_email(emailID)
+            email = valid.email
+        except EmailNotValidError as e:
+            self.logger.error(f"Register User: invalid email {emailID}")
```

Comment:
We won't really want to log errors from the client as errors encountered in the server.

The log severity used when logging a message should reflect the behavior of the server, not the client.  So here the server is behaving without any errors, it has detected an invalid email and reporting that back as a `400` error to the client.  So `INFO` is sufficient.  I like to use `WARNING` to make it quick to see what is happening with clients.

Also, our logger adds the module and function which the log message was emitted, so you don't need to add "Register User: " prefixes to the log messages.

And we don't want to use f-strings with log messages, see below.

And finally, well over 10+ times do we see the pattern `self.logger.*(); abort()`.  It might be worth considering a helper function for this.

Reply:
> We won't really want to log errors from the client as errors encountered in the server.
> 
> The log severity used when logging a message should reflect the behavior of the server, not the client. So here the server is behaving without any errors, it has detected an invalid email and reporting that back as a `400` error to the client. So `INFO` is sufficient. I like to use `WARNING` to make it quick to see what is happening with clients.

If we want to designate WARNING as exclusively "Pbench reporting a user/input error" so it's never confusing, that's fine; there certainly would be some value to being able to search for them as a class, as long as we're consistent about it.

The other option would be to create a new log severity like `INPUT` or `USER`... or simply adopt a convention of reporting them as INFO severity with an initial tag like `INPUT` in the message.

> Also, our logger adds the module and function which the log message was emitted, so you don't need to add "Register User: " prefixes to the log messages.

Yes, this was picked up from the original code, and we've been mindlessly following the pattern. I suppose we ought to fix that!

> And we don't want to use f-strings with log messages, see below.
> 
> And finally, well over 10+ times do we see the pattern `self.logger.*(); abort()`. It might be worth considering a helper function for this.

The problem with factoring logging into a method is that you lose the point of origin.... although I suppose that the Python `Logger` provides a way to hack around that like Java does.

Reply:
I am fine with consistently using WARNING for exclusively "Pbench user/input error"